File: Leetcode/2058/2058.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
func nodesBetweenCriticalPoints(head *ListNode) []int {
	var arr []int
	node := head
	for node != nil {
		arr = append(arr, node.Val)
		node = node.Next
	}
	var vec[]int
	for i := 1; i < len(arr)-1; i++ {
		if (arr[i] > arr[i-1] && arr[i] > arr[i+1]) || (arr[i] < arr[i-1] && arr[i] < arr[i+1]) {
			vec = append(vec, i)
		}
	}
	var ans []int
	if len(vec) < 2 {
		ans = append(ans, -1)
		ans = append(ans, -1)
		return ans
	}
	sort.Ints(vec)
	minVal := math.MaxInt32
	for i := 1; i < len(vec); i++ {
		minVal = min(minVal, vec[i]-vec[i-1])
	}
	maxVal := vec[len(vec)-1] - vec[0]
	ans = append(ans, minVal)
	ans = append(ans, maxVal)
	return ans
}

File: Leetcode/2058/2058.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
constexpr int INF = 0x3f3f3f3f;

class Solution {
public:
    vector<int> nodesBetweenCriticalPoints(ListNode* head) {
        vector<int> vec;
        auto node = head;
        while (node) {
            vec.emplace_back(node->val);
            node = node->next;
        }
        int n = (int)vec.size();
        vector<int> A;
        for (int i = 1; i < n-1; ++i) {
            if ( (vec[i] > vec[i-1] && vec[i] > vec[i+1]) || (vec[i] < vec[i-1] && vec[i] < vec[i+1]) ) {
                A.emplace_back(i);
            }
        }
        if (A.size() < 2) {
            return vector<int>{-1,-1};
        }
        sort(A.begin(),A.end());
        int maxDist = -1;
        int minDist = INF;
        int length = (int)A.size();
        if (length >= 2) {
            maxDist = max(maxDist, A.back() - A.front());
        }
        for (int i = 0; i < length-1; ++i) {
            minDist = min(minDist, A[i+1]-A[i]);
        }
        return vector<int>{minDist, maxDist};
    }
};